File: src/minidbms/storage/buffer.py
"""Fixed-capacity page cache with LRU and FIFO replacement."""

from __future__ import annotations

from collections import OrderedDict
from enum import Enum
import logging
from threading import RLock

from .errors import StorageError
from .file_manager import FileManager
from .page import PageFrame


logger = logging.getLogger(__name__)


class ReplacementPolicy(str, Enum):
    LRU = "LRU"
    FIFO = "FIFO"


class BufferPool:
    """Cache pages and write dirty frames back before eviction."""

    def __init__(
        self,
        capacity: int,
        policy: ReplacementPolicy | str,
        file_manager: FileManager,
    ) -> None:
        if isinstance(capacity, bool) or not isinstance(capacity, int) or capacity <= 0:
            raise StorageError("INVALID_BUFFER_CAPACITY", "buffer capacity must be positive")
        try:
            normalized_policy = (
                policy.value
                if isinstance(policy, ReplacementPolicy)
                else str(policy).upper()
            )
            self.policy = ReplacementPolicy(normalized_policy)
        except ValueError as error:
            raise StorageError(
                "UNKNOWN_REPLACEMENT_POLICY",
                f"replacement policy must be LRU or FIFO, got {policy!r}",
            ) from error

        self.capacity = capacity
        self.file_manager = file_manager
        self._frames: OrderedDict[int, PageFrame] = OrderedDict()
        self._lock = RLock()
        self._stats = {
            "read_requests": 0,
            "cache_hits": 0,
            "cache_misses": 0,
            "evictions": 0,
            "dirty_writes": 0,
        }

    def read_page(self, page_id: int) -> bytes:
        """Read a page through the cache and return an immutable copy."""
        with self._lock:
            self._stats["read_requests"] += 1
            frame = self._get_frame(page_id, record_cache_result=True)
            return bytes(frame.data)

    def write_page(self, page_id: int, data: bytes) -> None:
        """Update a cached page and mark it dirty without immediate disk I/O."""
        normalized = FileManager.normalize_page_data(data)
        with self._lock:
            frame = self._get_frame(page_id, record_cache_result=False)
            frame.data[:] = normalized
            frame.dirty = True
            self._touch(page_id)
            logger.debug("buffer write page=%s dirty=true", page_id)

    def pin_page(self, page_id: int) -> bytes:
        """Pin a page against eviction and return an immutable data copy."""
        with self._lock:
            self._stats["read_requests"] += 1
            frame = self._get_frame(page_id, record_cache_result=True)
            frame.pin_count += 1
            logger.debug("buffer pin page=%s count=%s", page_id, frame.pin_count)
            return bytes(frame.data)

    def unpin_page(self, page_id: int) -> None:
        with self._lock:
            frame = self._frames.get(page_id)
            if frame is None:
                raise StorageError("PAGE_NOT_CACHED", f"page {page_id} is not cached")
            if frame.pin_count == 0:
                raise StorageError("PAGE_NOT_PINNED", f"page {page_id} is not pinned")
            frame.pin_count -= 1
            logger.debug("buffer unpin page=%s count=%s", page_id, frame.pin_count)

    def free_page(self, page_id: int) -> None:
        """Release a page and remove any unpinned cached copy."""
        with self._lock:
            frame = self._frames.get(page_id)
            if frame is not None and frame.pin_count:
                raise StorageError(
                    "PAGE_PINNED", f"cannot free pinned page {page_id}"
                )
            self.file_manager.free_page(page_id)
            self._frames.pop(page_id, None)
            logger.debug("buffer discard freed page=%s", page_id)

    def flush_page(self, page_id: int) -> None:
        """Write one dirty cached page to disk."""
        with self._lock:
            frame = self._frames.get(page_id)
            if frame is None:
                # Validate the id. An allocated non-resident page is already clean.
                self.file_manager.read_page(page_id)
                return
            self._flush_frame(frame)

    def flush_all(self) -> None:
        """Write all dirty cached pages to disk."""
        with self._lock:
            for frame in self._frames.values():
                self._flush_frame(frame)

    def stats(self) -> dict[str, int]:
        """Return a copy so callers cannot mutate internal counters."""
        with self._lock:
            return dict(self._stats)

    def resident_page_ids(self) -> tuple[int, ...]:
        """Expose resident ids in replacement order for diagnostics/tests."""
        with self._lock:
            return tuple(self._frames)

    def _get_frame(self, page_id: int, *, record_cache_result: bool) -> PageFrame:
        frame = self._frames.get(page_id)
        if frame is not None:
            if record_cache_result:
                self._stats["cache_hits"] += 1
            self._touch(page_id)
            logger.info("buffer hit page=%s", page_id)
            return frame

        if record_cache_result:
            self._stats["cache_misses"] += 1
        logger.info("buffer miss page=%s", page_id)

        # Read before evicting so an invalid id cannot disturb a valid cache.
        page_data = self.file_manager.read_page(page_id)
        if len(self._frames) >= self.capacity:
            self._evict_one()
        frame = PageFrame(page_id=page_id, data=bytearray(page_data))
        self._frames[page_id] = frame
        return frame

    def _touch(self, page_id: int) -> None:
        if self.policy == ReplacementPolicy.LRU:
            self._frames.move_to_end(page_id)

    def _evict_one(self) -> None:
        victim_id = next(
            (
                page_id
                for page_id, frame in self._frames.items()
                if frame.pin_count == 0
            ),
            None,
        )
        if victim_id is None:
            raise StorageError(
                "NO_EVICTABLE_PAGE", "all cached pages are pinned"
            )
        victim = self._frames[victim_id]
        was_dirty = victim.dirty
        self._flush_frame(victim)
        del self._frames[victim_id]
        self._stats["evictions"] += 1
        logger.info(
            "buffer evict page=%s policy=%s dirty=%s",
            victim_id,
            self.policy.value,
            was_dirty,
        )

    def _flush_frame(self, frame: PageFrame) -> None:
        if not frame.dirty:
            return
        self.file_manager.write_page(frame.page_id, bytes(frame.data))
        frame.dirty = False
        self._stats["dirty_writes"] += 1
        logger.info("buffer flush dirty page=%s", frame.page_id)

File: cpp/src/storage/buffer.cpp
#include "wzt/buffer.hpp"

#include "wzt/errors.hpp"

#include <algorithm>
#include <cctype>

namespace wzt {

ReplacementPolicy parse_policy(std::string policy) {
    std::transform(policy.begin(), policy.end(), policy.begin(), [](unsigned char value) {
        return static_cast<char>(std::toupper(value));
    });
    if (policy == "LRU") return ReplacementPolicy::LRU;
    if (policy == "FIFO") return ReplacementPolicy::FIFO;
    throw StorageError("UNKNOWN_REPLACEMENT_POLICY", "replacement policy must be LRU or FIFO");
}

std::string policy_name(ReplacementPolicy policy) {
    return policy == ReplacementPolicy::LRU ? "LRU" : "FIFO";
}

BufferPool::BufferPool(std::size_t capacity, ReplacementPolicy policy, FileManager& file_manager)
    : capacity_(capacity), policy_(policy), file_manager_(file_manager) {
    if (capacity_ == 0) {
        throw StorageError("INVALID_BUFFER_CAPACITY", "buffer capacity must be positive");
    }
}

BufferPool::BufferPool(std::size_t capacity, const std::string& policy, FileManager& file_manager)
    : BufferPool(capacity, parse_policy(policy), file_manager) {}

PageData BufferPool::read_page(PageId page_id) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    ++stats_.read_requests;
    return get_frame(page_id, true).data;
}

void BufferPool::write_page(PageId page_id, const Bytes& data) {
    const auto normalized = FileManager::normalize_page_data(data);
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    auto& frame = get_frame(page_id, false);
    frame.data = normalized;
    frame.dirty = true;
    touch(page_id);
}

PageData BufferPool::pin_page(PageId page_id) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    ++stats_.read_requests;
    auto& frame = get_frame(page_id, true);
    ++frame.pin_count;
    return frame.data;
}

void BufferPool::unpin_page(PageId page_id) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    const auto found = frames_.find(page_id);
    if (found == frames_.end()) throw StorageError("PAGE_NOT_CACHED", "page is not cached");
    if (found->second.frame.pin_count == 0) throw StorageError("PAGE_NOT_PINNED", "page is not pinned");
    --found->second.frame.pin_count;
}

void BufferPool::free_page(PageId page_id) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    const auto found = frames_.find(page_id);
    if (found != frames_.end() && found->second.frame.pin_count != 0) {
        throw StorageError("PAGE_PINNED", "cannot free a pinned page");
    }
    file_manager_.free_page(page_id);
    if (found != frames_.end()) {
        order_.erase(found->second.order_position);
        frames_.erase(found);
    }
}

void BufferPool::flush_page(PageId page_id) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    const auto found = frames_.find(page_id);
    if (found == frames_.end()) {
        static_cast<void>(file_manager_.read_page(page_id));
        return;
    }
    flush_frame(found->second.frame);
}

void BufferPool::flush_all() {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    for (auto& item : frames_) flush_frame(item.second.frame);
}

BufferStats BufferPool::stats() const {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    return stats_;
}

std::vector<PageId> BufferPool::resident_page_ids() const {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    return {order_.begin(), order_.end()};
}

void BufferPool::set_log_stream(std::ostream* output) {
    std::lock_guard<std::recursive_mutex> guard(mutex_);
    log_output_ = output;
}

PageFrame& BufferPool::get_frame(PageId page_id, bool record_cache_result) {
    const auto found = frames_.find(page_id);
    if (found != frames_.end()) {
        if (record_cache_result) ++stats_.cache_hits;
        touch(page_id);
        log("buffer hit page=" + std::to_string(page_id));
        return found->second.frame;
    }
    if (record_cache_result) ++stats_.cache_misses;
    log("buffer miss page=" + std::to_string(page_id));
    const auto page_data = file_manager_.read_page(page_id);
    if (frames_.size() >= capacity_) evict_one();
    order_.push_back(page_id);
    const auto position = std::prev(order_.end());
    PageFrame frame{page_id, page_data, false, 0};
    auto inserted = frames_.emplace(page_id, CacheEntry{std::move(frame), position});
    if (!inserted.second) throw StorageError("IO_ERROR", "failed to insert buffer frame");
    return inserted.first->second.frame;
}

void BufferPool::touch(PageId page_id) {
    if (policy_ != ReplacementPolicy::LRU) return;
    auto& entry = frames_.at(page_id);
    order_.splice(order_.end(), order_, entry.order_position);
    entry.order_position = std::prev(order_.end());
}

void BufferPool::evict_one() {
    auto victim = order_.end();
    for (auto it = order_.begin(); it != order_.end(); ++it) {
        if (frames_.at(*it).frame.pin_count == 0) {
            victim = it;
            break;
        }
    }
    if (victim == order_.end()) throw StorageError("NO_EVICTABLE_PAGE", "all cached pages are pinned");
    const PageId victim_id = *victim;
    auto& frame = frames_.at(victim_id).frame;
    const bool dirty = frame.dirty;
    flush_frame(frame);
    frames_.erase(victim_id);
    order_.erase(victim);
    ++stats_.evictions;
    log("buffer evict page=" + std::to_string(victim_id) + " policy=" + policy_name(policy_) +
        " dirty=" + (dirty ? "true" : "false"));
}

void BufferPool::flush_frame(PageFrame& frame) {
    if (!frame.dirty) return;
    file_manager_.write_page(frame.page_id, Bytes(frame.data.begin(), frame.data.end()));
    frame.dirty = false;
    ++stats_.dirty_writes;
    log("buffer flush dirty page=" + std::to_string(frame.page_id));
}

void BufferPool::log(const std::string& message) const {
    if (log_output_ != nullptr) *log_output_ << message << '\n';
}

}  // namespace wzt
